### ISSUE TITLE
Resolve unclosed delimiter

### ISSUE THREAD

**Issue by amyjko** [wordplaydev/wordplay#334]
## What's the problem?

When an expression has an unclosed delimiter, it's annoying to have to type it:

`[ 1 2 3 4`

## What's the design idea?

Add a conflict resolution for `UnclosedDelimiter` that appends the missing delimiter.

## Who benefits?

Anyone looking for a quick fix to the problem.

## Design specification

Approved design: https://github.com/wordplaydev/wordplay/issues/334#issuecomment-2723683802

**Comment by yiqianna** [comment]
Hi! @amyjko Can I be assigned to this issue? Thank you.

**Comment by amyjko** [comment]
Sure! That's a lot of issues, are you sure you can complete them all?

**Comment by yiqianna** [comment]
Do you mean by there are lots of things to do with this issue? If so, would you mind explain them in more details? Thank you!

**Comment by amyjko** [comment]
Oh, I just meant you volunteered for a large number of issues, not on this issue in particular. When you're ready to get started on this one, just let me know, and I can post some details on how to get started on it.

**Comment by yiqianna** [comment]
Ohhh you probably mixed me up with someone else. This is the second issue I am assigned to and I already completed the first one. I am ready to work on this one.

**Comment by amyjko** [comment]
Ah, sorry about that, usernames are tricky to track. Let me know how I can help with this one.

**Comment by yiqianna** [comment]
### Proposal: Automatic Detection and Fix for Half-Deleted Delimiters in Wordplay Editor

### **1. Overview**
The Wordplay editor already features automatic delimiter pairing, where typing an opening bracket { [ ( automatically inserts the corresponding closing bracket } ] ). However, if a user manually deletes one half of the pair, the remaining unmatched delimiter creates an error. This disrupts code structure and requires manual correction.
This proposal introduces automatic detection and resolution for half-deleted delimiters, ensuring that users can quickly restore missing brackets without disrupting their workflow.

### **2. Problem Statement**
**Current Issues:**

- The editor automatically pairs brackets, but users can delete one side manually, breaking the structure.

- This leads to unbalanced expressions, requiring users to manually reinsert the missing delimiter.

- Users with motor impairments or those relying on screen readers may struggle to locate and correct missing delimiters.

- No auto-fix mechanism exists to detect and suggest restoring missing brackets.

### **3. Solution: Auto-Fix for Half-Deleted Delimiters**

1. **Detecting Half-Deleted Delimiters**

The editor will monitor deletions and identify cases where:

- A closing bracket () ] }) exists without a corresponding opening bracket.

- An opening bracket (( [ {) exists without a corresponding closing bracket.

2. **Providing Contextual Fix Suggestions**

When a mismatch is detected:

- A warning message will appear in the information panel, stating:

_"Unmatched delimiter detected. Would you like to restore the missing bracket?"_

- Provide an accessible “Restore Bracket” button, ensuring users can quickly apply the fix.

**3. Automatic Bracket Restoration**

If the user selects "Restore Bracket", the system will:

- Reinsert the missing delimiter at the correct position.

- Ensure the restored bracket matches the existing half.

**4. Keyboard Shortcut for Quick Fix**

For users who rely on keyboard navigation, a dedicated shortcut (e.g., Enter or Return) will allow them to restore a missing bracket without using a mouse.

### **4. Accessibility Considerations**

**1. Keyboard Navigation Support**

- Users should be able to navigate to the error message using the Tab key.

- The “Restore Bracket” button will be fully accessible via keyboard controls.

- The shortcut key (Enter or Return) will allow quick fixes without requiring a mouse.

**2. Screen Reader Compatibility**

- When an unmatched delimiter is detected, screen readers will announce:
_"Unmatched delimiter detected. Press Tab to navigate to the fix button, or use Enter or Return to restore."_

- The user presses Tab, and focus moves to the "Restore Bracket" button.

- The screen reader reads aloud:

_"Button: Restore Bracket. Press Enter to activate."_

- When a bracket is automatically restored, the screen reader will confirm:

_"Bracket restored at position X."_

**3. Alternative Input Compatibility**

- Users with motor impairments who use voice commands (e.g., speech-to-text) should be able to say:

_“Restore missing bracket”_ → Triggers the fix.

- The feature should integrate with assistive technologies, such as switch devices or eye-tracking input systems.

### **5. Implementation Roadmap**

**Phase 1: Detection & Contextual Suggestions**

- Implement real-time detection of unmatched brackets.

- Display error messages and fix suggestions in the information panel.

- Ensure basic screen reader support for error notifications.

**Phase 2: Automated Correction & Keyboard Shortcuts**

- Implement the “Restore Bracket” button functionality.

- Introduce keyboard shortcuts for quick fixes.

**Phase 3: Accessibility Refinements & Testing**

- Conduct usability testing with keyboard-only users.

- Ensure compatibility with screen readers and alternative input devices.

- Refine spoken feedback and voice command support.

### **6. Expected Outcomes**

**Seamless Editing Experience** – Users no longer need to manually reinsert missing brackets.
**Reduced Syntax Errors** – Automatically fixes unbalanced expressions before they cause issues.
 **Full Keyboard and Screen Reader Support** – Ensures accessibility for users with disabilities.
**Improved Visual Cues** – Helps users easily detect missing delimiters.

**Comment by amyjko** [comment]
This is a good start! A few notes, to help simplify it:

* There are many other auto-correction features implemented, and this should generally follow the same pattern, for consistency. To see examples, try typing your phone number, or putting a typo in reference to a name.
* The compiler already detects unclosed delimiters; write a program with a missing one and you will see the message. The part of the design that proposes to detect and warn about them isn't needed. The design should build upon the existing conflict messages that are already present.
* "Restore bracket" isn't quite accurate; that language implies it was removed. "Add missing [insert symbol name here]" would be more accurate.
* We don't yet have speech recognition in the system. But the design mentioned is already covered by operating system level accessibility features, which allow for speaking standard WCAG compliant button labels.
* There's no need for an implementation roadmap, just a design proposal. (Unless you also want to be the developer for this. If you do, I'm happy to explain the patterns we use to implement auto-correction features).

**Comment by yiqianna** [comment]
### **Proposal: Automatic Detection and Fix for Half-Deleted Delimiters in Wordplay**
**1. Overview**
The Wordplay editor already features automatic delimiter pairing, where typing an opening bracket { [ ( automatically inserts the corresponding closing bracket } ] ). However, if a user manually deletes one half of the pair, the remaining unmatched delimiter creates an error. This disrupts code structure and requires manual correction.
This proposal builds on the existing conflict resolution system by introducing an auto-correction option that allows users to quickly insert the missing delimiter using a single keypress or button press.
**2. Consistency with Existing Auto-Correction Features**

- This feature should follow the same auto-correction flow as existing Wordplay corrections (e.g., fixing typos or formatting phone numbers).

- The existing compiler conflict message for unclosed delimiters will be extended to offer an auto-correction action.

**3. User Interaction Flow**

1. User manually deletes a delimiter (e.g., deletes ] from [1 2 3 4 ]).

2. The existing conflict message appears:
"Missing delimiter detected: Add missing ]?"

3. User Options:
Press Enter → Auto-corrects the missing delimiter (fastest option).
Click "Add Missing ]" Button → Restores the delimiter manually.
Ignore → The message stays on when the user continues typing.

**Comment by amyjko** [comment]
I think this is pretty close, with two remaining issues:

* How would the editor distinguish between `Enter` to create a new line and `Enter` to  add the missing delimiter? Tabbing into the conflict, and onto the button would make sense, but there's no way the editor would be able to distinguish the two.
* I'm not sure what ignore is supposed to do. Ignoring the warning forever would be very unhelpful, as it's a real syntax error that prevents accurate interpretation of the intention of the creator. The conflict message is already ignorable.

**Comment by yiqianna** [comment]
### **Proposal: Automatic Detection and Fix for Half-Deleted Delimiters in Wordplay**

**1. Overview**
The Wordplay editor already features automatic delimiter pairing, where typing an opening bracket { [ ( automatically inserts the corresponding closing bracket } ] ). However, if a user manually deletes one half of the pair, the remaining unmatched delimiter creates an error. This disrupts code structure and requires manual correction.
This proposal builds on the existing conflict resolution system by introducing an auto-correction option that allows users to quickly insert the missing delimiter using a single keypress or button press.

**2. User Interaction Flow**

1. User manually deletes a delimiter (e.g., deletes ] from [1 2 3 4 ]).

2. The existing conflict message appears:"Missing delimiter detected: Add missing ]?"

3. User Options:
- Press Tab to navigate to the "Add Missing ]" button → Press Enter to confirm the fix.
- Click "Add Missing ]" Button → Restores the delimiter manually.
- Continue typing → The message remains until the delimiter is corrected; it does not disappear unless the issue is resolved.

**Comment by amyjko** [comment]
Thanks @yiqianna, this simple design will suffice. Design approved!

**Comment by NatnaelGoit** [comment]
Hello @amyjko may I be assigned to this issue?

**Comment by NatnaelGoit** [comment]
Thank you!

**Comment by NatnaelGoit** [comment]
Yes I'm still working on this issue sorry for the late response, I plan on having this done before winter quarter starts though so I'll make sure to keep you posted!

**Comment by amyjko** [comment]
Thanks for the update! Let me know how I can help.

**Comment by amyjko** [comment]
Resolved by 5631990ae47157bed1753eb7608cffa3676cc337.
